### PR TITLE
Run OnionService with Tor in a dedicated process

### PR DESCRIPTION
**Pull request by grote**

### CODE REVIEW

File: app/src/main/java/org/onionshare/android/tor/TorManager.kt [hunk 1]
```diff
@@ -45,46 +45,41 @@ private val EVENTS = listOf(
...
+                            } catch (e: Exception) {
+                                // gets caught and logged by caller
+                                continuation.resumeWithException(e)
+                                return
+                            }
+                            val onion = createOnionService(this, continuation, port)
+                            // if onion is null, we resume with an exception in createOnionService()
+                            if (onion != null) addRawEventListener { keyword, data ->
                                 LOG.debug("$keyword: $data")
+                                if (keyword == "HS_DESC" && data.startsWith("UPLOADED ")) {
```
Comment (akwizgran) [inline app/src/main/java/org/onionshare/android/tor/TorManager.kt:78]: Might be useful to assert that the message contains the right onion address, to catch lifecycle bugs?

File: app/src/main/java/org/onionshare/android/tor/TorManager.kt [hunk 2]
```diff
@@ -45,46 +45,41 @@ private val EVENTS = listOf(
...
+                                    if (continuation.isActive) continuation.resume(onion)
```
Comment (akwizgran) [inline app/src/main/java/org/onionshare/android/tor/TorManager.kt:79]: Should we remove the listener after the first upload event?

File: app/src/main/java/org/onionshare/android/tor/TorManager.kt [hunk 3]
```diff
@@ -125,24 +114,48 @@ class TorManager @Inject constructor(
...
+        val controlFileDescriptor = client.fileDescriptor
+        val inputStream = FileInputStream(controlFileDescriptor)
+        val outputStream = FileOutputStream(controlFileDescriptor)
+        return TorControlConnection(inputStream, outputStream)
+    }
+
+    /**
+     * Returns the absolute path to the control socket on the local filesystem.
+     *
+     * Note: Exposing this through TorService was rejected by upstream.
```
Comment (akwizgran) [inline app/src/main/java/org/onionshare/android/tor/TorManager.kt:152]: :dagger:

**Review by akwizgran — APPROVED** [review]
Looks good!

File: app/src/main/java/org/onionshare/android/tor/TorManager.kt (hunk 2 above)
Comment (grote) [reply app/src/main/java/org/onionshare/android/tor/TorManager.kt:79]: Depends. If we don't want/need to see the Tor logs, we could.

File: app/src/main/java/org/onionshare/android/tor/TorManager.kt (hunk 1 above)
Comment (grote) [reply app/src/main/java/org/onionshare/android/tor/TorManager.kt]: Good point, I pushed  a new commit that checks for the right address (and then removes the listener).

File: app/src/main/java/org/onionshare/android/tor/TorManager.kt (hunk 2 above)
Comment (akwizgran) [reply app/src/main/java/org/onionshare/android/tor/TorManager.kt]: Is it a problem if the listener keeps a reference to the continuation and keeps calling resume() on it after the continuation would normally have finished? Hopefully that's not a memory leak?

Comment (akwizgran) [reply app/src/main/java/org/onionshare/android/tor/TorManager.kt]: Oh sorry, just saw in the other thread that you're removing the listener now. Never mind.

**Pull request merged**